### PR TITLE
Added focus-changed-event callback to FocusScope

### PR DESCRIPTION
**Pull request by FloVanGH**
```slint
FocusScope {
    focus-changed-event => {
        debug("Focus has changed");
    }
}
```

### CODE REVIEW

**Review by ogoffart — COMMENTED** [review]
Nice!

Missing documentation and tests.

And also we probably should discuss the naming.

It is called `key_pressed`  so i think it should be called  `focussed`  and `focus-lost`
Or something maybe `focus-gained` ?

**Comment by FloVanGH** [comment]
> `focussed`

Yes I also thought about ` focussed`.

**Comment by FloVanGH** [comment]
> Nice!
> 
> Missing documentation and tests.
> 
> And also we probably should discuss the naming.
> 
> It is called `key_pressed` so i think it should be called `focussed` and `focus-lost` Or something maybe `focus-gained` ?

@tronical what is your favorite naming ;-) .

**Comment by ogoffart** [comment]
I wonder if focused-gained is better naming than focused.    @simon:  what do you think?

**Comment by ogoffart** [comment]
I meant @tronical   (sorry @ simon, whoever you are :-) )

**Comment by tronical** [comment]
If we call it `focused` then the counter-part can't be `focus-lost` (not symmetric) but I think it should be `unfocused`.

I agree that `focus-gained` and `focus-lost` is the other symmetric option. That said, I'm not so much of a fan of "gain", the is a term that's used with a different meaning in electrical engineering for example. A synonymous `focus-obtained` would be my preference, if we go for `focus-obtained` / `focus-lost`.

On the web the DOM callbacks are called `focusin` and `focusout`, that would be `focus-in` and `focus-out` in our "world" I think. But the web also has the "blurred" and "focused" concept as counter-parts. I find "blurring" confusing personally, but it would be a consistent choice for us that maintains the style (`XXXed`): `focused` and `blurred`.

On that note: If we go for `focused` then it should be `focused`, not `focussed` - the latter is AFAIK old English spelling, we want to pick the modern English variant I'd say.

But to be perfectly honest, I think these focus callbacks are _not_ like `clicked`, or `moved` in `TouchArea`. I much prefer the `-event` suffix naming, such as `scroll-event`, `pointer-event`. I wish we had call it `key-pressed-event` and `key-released-event` in `FocusScope`, but instead it's currently called `key-pressed` and `key-released`.

If we want to continue with the `-event` names, then my preference here would be a `focus-event` with a struct and a `focused` or `has-focus` boolean field. That way we can also add more information later that we'll need, for example temporary loss of focus due to a popup.

Finally, as the property change handlers are "looming" on the horizon: Wouldn't it be called `has-focus-changed` automatically in the future? Should we call it like that now perhaps? Or should we simply go with the one correct name that's simple and concise: `focus-changed`?

So my preference here is:

1. `focus-event` with a struct that has a boolean and is extensible for focus change reasons in the future.
2. `focus-changed` and in the handler you can check the `has-focus` property.
3. `has-focus-changed` (because we're going to get that in the future anyway.... right? :)
4. `focused` and `blurred`
5. `focused` and `unfocused`

2-5 suffer from the missing extensibility. That's why 1 is my preferred option, but (2) seems also very elegant to me.

**Comment by tronical** [comment]
I'm wondering what @jpnurmi 's take on this is :)

**Comment by FloVanGH** [comment]
> 1. focus-event with a struct that has a boolean and is extensible for focus change reasons in the future.

What extra fields could it be, in the future. If there is never a change that we add more stuff I think a struct for this is a little bit overkill.

> 2. focus-changed and in the handler you can check the has-focus property.

I think in common we should prevent to add names like `xxx-changed` because if we will introduce property changed callbacks later, me could end up in a situation with name collisions or if not it could be still confusing.

>  3. has-focus-changed (because we're going to get that in the future anyway.... right? :)

Same as 2.

If we will have a situation in the future that more fields are necessary I would prefer 1. otherwise 5.

**Comment by tronical** [comment]
> > 1. focus-event with a struct that has a boolean and is extensible for focus change reasons in the future.
> 
> What extra fields could it be, in the future. If there is never a change that we add more stuff I think a struct for this is a little bit overkill.

The only field I can think of at this moment is the reason for the focus change. For example if you have a `TextInput` that draws a focus border around its text edit, and the `TextInput` looses focus, that border should go away. But that's not the case when the focus is temporarily lost due to a popup being shown _for_ the very text input (the type of popup that offers copy / paste, etc.). This is why Qt has a focus reason associated with its focus events.

> 
> > 2. focus-changed and in the handler you can check the has-focus property.
> 
> I think in common we should prevent to add names like `xxx-changed` because if we will introduce property changed callbacks later, me could end up in a situation with name collisions or if not it could be still confusing.

I think either way we're going to end up with a duplicate that we might deprecate. If you leave out the property change events, etc. parts, isn't `focus-changed` much better than _two_ callbacks for just one boolean?

> > 3. has-focus-changed (because we're going to get that in the future anyway.... right? :)
> 
> Same as 2.

Well, if that _is_ going to be the name in the future anyway, then doesn't it make sense to introduce it now instead of introducing a _second_ name (5.) that's already obsolete on introduction?

**Comment by ogoffart** [comment]
I'm not a fan of a single focus-event. for two reason:
 
1. the thing we do when getting and losing the focus are different.  In a design tool for example i would like to "link" focus-gained to some other callback or action, and i want to split between the gained and the lost.

2. when we get property changed callback it becomes redundant.


IMHO, `focus-gained` is quite obvious. Focus is "gained", says my language model.
Nobody would be confused about the use in electrical engeneering

That's the api from Java's FocusListener  https://docs.oracle.com/javase/8/docs/api/java/awt/event/FocusListener.html

The QML has a single onFocusChanged  , but the documentation clearly talk about "gain" / "gained" https://doc.qt.io/qt-6/qml-qtquick-item.html#focus-prop (also activeFocus prop)

The flutter docs also has a single callback, and the documentation also talk about gaining and losing focus https://api.flutter.dev/flutter/widgets/Focus/onFocusChange.html

In many stack overflow question i found on google when searching for focus change event in different framework, many of them use the term "gain" focus.

so personally, i'd go for `focus-gained`.

**Comment by jpnurmi** [comment]
If the simple boolean property change handler is going to be there later anyway, then I'd vote for the low-level `-event` variant(s) with access to the event and focus reason to allow more complex widgets to adapt their behavior depending on the focus reason.

For styling purposes, having a focus reason property in `FocusScope` would make it easy to use declarative bindings for visualizing click vs. tab focus differently.

**Comment by ogoffart** [comment]
More data points: 

Winit uses `Focused(bool)`  when  "The window gained or lost focus"
https://docs.rs/winit/latest/winit/event/enum.WindowEvent.html#variant.Focused

egui:  `gained_focus` / `lost_focus`  https://docs.rs/egui/latest/egui/struct.Response.html#method.gained_focus

iced:  has [`Focused`](https://docs.rs/iced/latest/iced/window/enum.Event.html#variant.Focused) and `Unfocused` events  (there is a [`gain_focus` function](https://docs.rs/iced/latest/iced/window/fn.gain_focus.html) but that's unrelated 

dioxus:  `onfocus` `onfocusin` `onfocusout`  (but they are really close to web stuff) https://docs.rs/dioxus/latest/dioxus/events/index.html?search=onfocus

trauri:  same as winit https://docs.rs/tauri/latest/tauri/enum.WindowEvent.html#variant.Focused

druid: [FocusChanged](https://docs.rs/druid/latest/druid/widget/prelude/enum.LifeCycle.html#variant.FocusChanged)  the docs talk about "gain" as well.

**Comment by tronical** [comment]
Afaics this isn't just about adding two callbacks to FocusScope but users will ask for them to be added to any widget that has a `has-focus` property (TextEdit, LineEdit, ComboBox). We had the same situation with the font properties, they are transitive.

Then in the future they will either have both (has-focus-changed and these two callbacks), or we will decline and say only FocusScope will have the callbacks, everyone else must use the change handler. That feels inconsistent.

The more I think about it, the more I feel that this is really a change handler and that's how the API should be like. It's not the same as the clicked callback.

At best an -event, but at heart a change callback.

**Comment by ogoffart** [comment]
> only FocusScope will have the callbacks,

That's how I see it.
These callback are to implement behavior when you implement your own widget.

But now that you write your message, i can indeed see the need that user will have to detect that in all the widgets. And indeed that's not great.

> added to any widget

(Note: maybe we really need a `Widget` base that has some common properties common for all widgets, but's that's another topic)

> At best an -event, but at heart a change callback.

I'm still much in favor of `focus-gained` and `focus-lost`.
but if you insist on being a event, then its better than nothing.

Note that the change callback, contrary to the direct callback, might be called in a later iteration of the event loop.

**Comment by FloVanGH** [comment]
After reading your feedback. Maybe it's a better step to start with change handlers first. It would give each user the ability to do and call stuff after a property is changed. With that not only `FocusScope` would profit from the `has-focus-changed`, also `LineEdit`, `TextEdit`, ... . And then as next step if needed we can to think about specific callbacks e.g.`focus-gained` or `focus-event` on `FocusScope`.

**Comment by FloVanGH** [comment]
@tronical I checked how to implement a  `focus-event` callback on the `FocuScope` I've noticed that we already have a  `FocusEvent` type internal but I think that is not the thing what we want to use. 

What we want is something is a `FocusChangeEvent`, with a `focused: bool` field right? 

An other question, should the user be able to `reject` the `focus-event` like in `key-pressed`? If yes then in the focus scope I first need to check the result of the callback and only if it is accepted to set `has-focus` to true. And if it is rejected return `FocusEventResult::FocusIgnored` from  `focus_event` on the rust `FocusScope` right?

If we want to implement it as focus event maybe `focus-change-event` would be a better name, because we already have a  `focus_event` on rust. It's not visible in rust but nerveless I think it make sense to give them both different names.

**Comment by tronical** [comment]
> @tronical I checked how to implement a `focus-event` callback on the `FocuScope` I've noticed that we already have a `FocusEvent` type internal but I think that is not the thing what we want to use.
>
> What we want is something is a `FocusChangeEvent`, with a `focused: bool` field right?
> 
> An other question, should the user be able to `reject` the `focus-event` like in `key-pressed`? If yes then in the focus scope I first need to check the result of the callback and only if it is accepted to set `has-focus` to true. And if it is rejected return `FocusEventResult::FocusIgnored` from `focus_event` on the rust `FocusScope` right?
> 
> If we want to implement it as focus event maybe `focus-change-event` would be a better name, because we already have a `focus_event` on rust. It's not visible in rust but nerveless I think it make sense to give them both different names.

Since we're talking about a callback in `FocusScope` I don't think that we need to worry about accepting/ignoring the internal event. That is already dealt with by `FocusScope`'s `enabled` property, to which users can create bindings for.

I think `focus-change-event` is a good name, indeed a bit better than the more generic `focus-event`. Makes it clear that this isn't about accepting/ignoring, but merely a timely notification.

I don't think that the callback needs to accept any arguments, since the `FocusEvent`'s `has-focus` property will have the correct value, and as JP rightly pointed out, the focus change reason is something that should be a property in the future.

**Comment by ogoffart** [comment]
> because we already have a focus_event on rust. It's not visible in rust but nerveless I think it make sense to give them both different names.

Internal names is not a reason to give worse names in the public API.  If there is a name clash, we can just rename the internal event.

I suppose we could be able to reject the focus and that would be the point of it. But the default should be to accept, and unfortunately, the current EventResult defaults to reject.

We also should give a reason on why we get the focus (clicked, tab, programmatic, ??), but we can add that later
(IMHO I think for that use case, it'd be best to have a property on the FocusScope that tells if the FocusScope can take focus with clicks or with tabs, that's what would be in the design tool.)

**Comment by tronical** [comment]
> (IMHO I think for that use case, it'd be best to have a property on the FocusScope that tells if the FocusScope can take focus with clicks or with tabs, that's what would be in the design tool.)

👍 👍

File: docs/reference/src/language/builtins/elements.md [hunk 1]
```diff
@@ -212,6 +212,7 @@ or it will be mapped to a private unicode character. The mapping of these non-pr
     forwarded to the parent element.
 -   **`key-released(KeyEvent) -> EventResult`**: Invoked when a key is released, the argument is a [`KeyEvent`](structs.md#keyevent) struct. The returned [`EventResult`](enums.md#eventresult) indicates whether to accept or ignore the event. Ignored events are
     forwarded to the parent element.
+-   **`focus-changed-event()`**: Invoked when the focus on the focus scope has changed.
```
Comment (tronical) [inline docs/reference/src/language/builtins/elements.md:215]:
```suggestion
-   **`focus-changed-event()`**: Invoked when the focus on the `FocusScope` has changed.
```

Alternate idea that makes it a little clearer what to do in the handler:

```suggestion
-   **`focus-changed-event()`**: Invoked when the `has-focus` property on the `FocusScope` has changed.
```

File: internal/compiler/builtins.slint [hunk 2]
```diff
@@ -109,6 +109,8 @@ export component FocusScope {
     out property <bool> has-focus;
     callback key_pressed(KeyEvent) -> EventResult;
     callback key_released(KeyEvent) -> EventResult;
+    callback focused();
```
Comment (tronical) [inline internal/compiler/builtins.slint:112]: This looks like a leftover?

**Review by tronical — COMMENTED** [review]
Looks overall good to me. A few nitpicks inside, but I like this API. It's minimal, and straight to the point.

File: internal/compiler/builtins.slint (hunk 2 above)
Comment (FloVanGH) [reply internal/compiler/builtins.slint]: Yes right. Is removed.

**Review by tronical — APPROVED** [review]
Looks good to me. Please adjust the title of this PR (and the description) to match the commits.

@ogoffart , what do you think? ?

**Review by ogoffart — APPROVED** [review]
(i would have preferred two callbacks, but this is fine.  One callback is less overhead anyway)

**Pull request merged**